### ISSUE TITLE
Langfuse integration

### ISSUE THREAD

**Issue by elijahbenizzy** [DAGWorks-Inc/burr#206]
**Is your feature request related to a problem? Please describe.**
We should be able to automatically track every Burr app using Langfuse automatically.

With hooks (https://burr.dagworks.io/concepts/hooks/) one should be able to do this nicely.

**Describe the solution you'd like**

A trace-based langfuse integration.  The following features at a minimum:
- [ ] One trace per application run (this is dependent on this: https://github.com/DAGWorks-Inc/burr/issues/203)
- [ ] One span per step
- [ ] One span per tracer span call
- [ ] Observations at will (perhaps state elements)
- [ ] Perhaps access to the langfuse client.
 
You'll want to follow the same classes as this does: https://github.com/DAGWorks-Inc/burr/blob/main/burr/tracking/client.py#L60,
except for the `BaseStateLoader`, that's not relevant. Then you can store state in it. 

See this for an inspiration (this is specific to datadog): https://github.com/DAGWorks-Inc/hamilton/blob/96da3107666dcccdd80da39d7d4d118b4de82571/hamilton/plugins/h_ddog.py#L19.

There are two ways to do this:
1. Expose the langfuse client in the functions/classes
2. Bury it behind the logger

We should enable both, but strive towards (2).

For (1), we can do it if you pass the langfuse object in for application calls as runtime inputs. This will make it easier: https://github.com/DAGWorks-Inc/burr/issues/205. 

For (2) we'll need to reconcile our model with theirs. Specifically, this will help: https://github.com/DAGWorks-Inc/burr/issues/46.

Let's do (1) then open up another issue for (2)

**Describe alternatives you've considered**

**Additional context**

**Comment by fer-git** [comment]
+1 built-in langfuse integration please.

**Comment by andreahlert** [comment]
Labels migrated to new taxonomy (#692).

**Comment by elijahbenizzy** [comment]
If anyone is excited about taking this on you are welcome! OTel should be good but we can probably go a bit further.

**Comment by hannguyen0712** [comment]
Hi @elijahbenizzy, I'd like to take this on! I'm looking to work on contributing to LLM-observability integration.
One scoping question first, since the issue has aged a bit. You noted on Mar 29 that "OTel should be good but we can probably go a bit further," and I see #205 (global inputs, which the original approach (1) depended on) was since closed as not planned. Given Langfuse now ingests OpenTelemetry natively, would you prefer I build this on top of Burr's OTel tracing support, with emitting spans Langfuse can consume, rather than a dedicated Langfuse client following the burr/tracking/client.py pattern? That seems like the cleaner path and avoids the dead #205 dependency, but I want to confirm before starting.
Assuming OTel is the direction, I'd scope a focused first PR via Burr's hooks: one trace per application run, one span per step, with state observations attached, plus a short example and docs. I can open follow-ups for anything beyond that (e.g. exposing the Langfuse client directly, or per-tracer-span observations). Does that work as a starting point?

**Comment by elijahbenizzy** [comment]
@hannguyen0712 looking forward to it! Can work both ways -- up to you. I think we can get around 205!

**Comment by hannguyen0712** [comment]
Working on a draft PR for this

**Comment by hannguyen0712** [comment]
I've implemented the feature and I'm working on some testing before submitting. My apologies for any delays - I took a week off for July 4th holiday.

**Comment by hannguyen0712** [comment]
Hi @elijahbenizzy, PR is up as #844 . I went with the OTel route we discussed. The Langfuse Python SDK is now v4, which filters span export to LLM-relevant spans by default and would silently drop Burr spans, so the bridge injects an export filter to handle it (details in the PR). Scope is what we agreed: trace per run, span per step and per tracer span, state observations, and the client exposed on the bridge, plus tests, docs, and an example verified end to end on Langfuse Cloud (screenshots in the PR).

As I was working on it, while testing error propagation, I found that post_run_execute_call hooks always receive exception=None (the wrapper never captures the raised exception, so even the existing OTel bridge can't mark failed runs as errored at the root). I have a small 8-line fix that touches core with regression tests ready. Would you want me to file it as an issue, a separate PR, or fold it into this one? 

The original issue also said to open a follow-up for approach (2). I'm happy to file that tracking issue too if you'd like. 

Would love your review! Thank you for your time!
